### DISCUSSION TITLE
Using PYMUPDF to reconstruct PDF, previewing in some front-end libraries may lose some display of vector images,

### DISCUSSION THREAD

**Discussion by K8S666** [pymupdf/PyMuPDF#3893] [Looking for help]
Using PYMU to reconstruct PDF, previewing in some front-end libraries may lose some display of vector images

Preconditions:
I tried using PYMU to extract the text and coordinates from the document, and replaced the new text with the original coordinates. After making modifications on the original PDF, I saved the document

result:
Preview the modified document using the front-end library and find that some vector lines or graphics on the page are not displayed; But when I preview the same document using Adobe PDF, it can still be fully displayed;

Question:
I'm not sure if it's my PYMU refactoring method that caused some graphics to be lost or something else;

The left side of the attached image shows the preview effect in the frontend library, the left side shows the source file, and the right side shows the processed file; Please also attach the files before and after processing

**Comment by K8S666** [comment]
The test found that the written text does not belong to any layer. This may be the reason why other tools fail to render. Is there a way to specify the text insertion layer?

**Reply by JorjMcKie** [reply]
You can use the `oc` parameter. This must be the xref of an OCG (roughl equivalent to "layer").

**Comment by JorjMcKie** [comment]
If you use redactions to replace text, please watch out how you set the `graphics` parameter.
By default, contained drawings will be removed.

**Reply by K8S666** [reply]
page.apply_redactions(images=False, graphics=fitz.PDF_REDACT_LINE_ART_NONE)

This is my parameter setting. How should I adjust it?

**Reply by JorjMcKie** [reply]
This does look good.

**Reply by K8S666** [reply]
Thank you for your answer. This question has been bothering me for a long time. My latest observation is that the instructions for vector graphics in the processed instruction stream of PDF have changed from 0.233 to. 233. This may directly result in some insufficiently robust parsing libraries being unable to parse the correct format. My question is, is the issue with this instruction stream caused by PYMUPDF writing
![image](https://github.com/user-attachments/assets/7795c571-c01e-4e9a-98fa-c662356c568b)

**Reply by K8S666** [reply]
![image](https://github.com/user-attachments/assets/cb916aec-778b-49be-b000-7a8881f09740)
I think I have encountered the same problem as the illustration, and I also think it may be a bug in the latest version
